applications/services/gui/modules/number_input.c

```diff
@@ -0,0 +1,493 @@
+ */
+static void number_input_draw_input(Canvas* canvas, NumberInputModel* model) {
+    const uint8_t text_x = 8;
+    const uint8_t text_y = 25;
+
+    elements_slightly_rounded_frame(canvas, 6, 14, 116, 15);
+
+    FuriString* text = model->text_buffer;
+    canvas_draw_str(canvas, text_x, text_y, furi_string_get_cstr(text));
+}
+
+static void number_input_backspace_cb(NumberInputModel* model) {
+    uint8_t text_length = model->clear_default_text ? 1 :
+                                                      furi_string_utf8_length(model->text_buffer);
+    if(text_length > 0) {
+        furi_string_set_strn(
+            model->text_buffer, furi_string_get_cstr(model->text_buffer), text_length - 1);
+    }
+}
+
+static char* int32_to_string(int32_t value) {
+    char* buffer = (char*)malloc(12);
+    snprintf(buffer, 12, "%ld", value);
+    return buffer;
+}
+
+/** Handle up button
+ *
+ * @param      model  The model
+ */
+static void number_input_handle_up(NumberInputModel* model) {
+    if(model->selected_row > 0) {
+        model->selected_row--;
+        if(model->selected_column > number_input_get_row_size(model->selected_row) - 1) {
+            model->selected_column = number_input_get_row_size(model->selected_row) - 1;
+        }
+    }
+}
+
+/** Handle down button
+ *
+ * @param      model  The model
+ */
+static void number_input_handle_down(NumberInputModel* model) {
+    if(model->selected_row < keyboard_row_count - 1) {
+        if(model->selected_column >= number_input_get_row_size(model->selected_row) - 1) {
+            model->selected_column = number_input_get_row_size(model->selected_row + 1) - 1;
+        }
+        model->selected_row += 1;
+    }
+    const NumberInputKey* keys = number_input_get_row(model->selected_row);
+    if(keys[model->selected_column].text == sign_symbol && !model->useSign) {
+        model->selected_column--;
+    }
+}
+
+/** Handle left button
+ *
+ * @param      model  The model
+ */
+static void number_input_handle_left(NumberInputModel* model) {
+    if(model->selected_column > 0) {
+        model->selected_column--;
+    } else {
+        model->selected_column = number_input_get_row_size(model->selected_row) - 1;
+    }
+    const NumberInputKey* keys = number_input_get_row(model->selected_row);
+    if(keys[model->selected_column].text == sign_symbol && !model->useSign) {
+        model->selected_column--;
+    }
+}
+
+/** Handle right button
+ *
+ * @param      model  The model
+ */
+static void number_input_handle_right(NumberInputModel* model) {
+    if(model->selected_column < number_input_get_row_size(model->selected_row) - 1) {
+        model->selected_column++;
+    } else {
+        model->selected_column = 0;
+    }
+    const NumberInputKey* keys = number_input_get_row(model->selected_row);
+    if(keys[model->selected_column].text == sign_symbol && !model->useSign) {
+        model->selected_column++;
+    }
+}
+
+static void prevent_to_large_number(NumberInputModel* model) {
```

Comment:
In terms of UI/UX this behavior is quite painful: it irreversibly modifies previous input.
Best option is to ignore input. Other systems also have visual/audial alert, but we don't use them for input.